### PR TITLE
feat(terraform): rework dep types

### LINKED ISSUE #8792: breaking: Rework depTypes of the Terraform manager
**What would you like Renovate to be able to do?**

To allow a more fine grained control over the behavior of the Terraform manager a rework of the used `depType` values is needed.    
For example it should be possible to deactivate the renovating of required providers or the required Terraform version. 

Context: https://github.com/renovatebot/renovate/issues/8746#issuecomment-781378142

**Did you already have any implementation ideas?**

If I understand the underlying mechanics correctly, `depType` has no longer any affects outside of text rendering  as long as a data source is provided. 

| ressource        | suggested depType           |  current value | datasources used |
| ------------- |:-------------:| :-------:|:-----:|
| terraform provider      | provider |  terraform  | terraform-provider |
| required terraform  provider     | required_provider      |  terraform | terraform-provider |
| required terraform version | required_version  |  terraform | github-tags |
| terraform module | module  | terraform, github, gitTags | terraform-module, github-tags, git-tags |
| helm release | helm_release  |  helm | helm |
| docker container | docker_container | stage, final (inherited from `Dockerfile` manager) | docker ( inherited ) |
| docker image | docker_image  | stage, final (inherited from `Dockerfile` manager) | docker ( inherited ) |
| docker service | docker_service  | stage, final (inherited from `Dockerfile` manager) | docker ( inherited ) |

### PR DESCRIPTION
**Pull request by secustor**
## Changes:

This PR changes all `depTypes` provided by the Terraform manger to allow a more fain grained control over which dependencies are updated and which are not. 

## Context:

Closes #8792

I'm not sure how to do the migration for users which use the `helm` depType.    
Can you point me to some example PR where this has already been done? 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/manager/terraform/readme.md [hunk 1]
```diff
@@ -1,19 +1,36 @@
...
 
 - `>= 1.2.0`: version 1.2.0 or newer
 - `<= 1.2.0`: version 1.2.0 or older
 - `~> 1.2.0`: any non-beta version >= 1.2.0 and < 1.3.0, e.g. 1.2.X
 - `~> 1.2`: any non-beta version >= 1.2.0 and < 2.0.0, e.g. 1.X.Y
 - `>= 1.0.0`, <= 2.0.0`: any version between 1.0.0 and 2.0.0 inclusive
+
+For fine-grained control, e.g. to turn off only parts of this manager, there are following `depTypes` provided:
+
+| ressource                   |      depType      |
```
Comment (HonkingGoose) [inline lib/manager/terraform/readme.md:27]:
```suggestion
| resource                   |      depType      |
```

Fix typo

File: lib/manager/terraform/readme.md [hunk 2]
```diff
@@ -1,19 +1,36 @@
-Currently Terraform support is limited to Terraform registry sources and github sources that include SemVer refs, e.g. like `github.com/hashicorp/example?ref=v1.0.0`.
+Currently, Terraform supports renovating following dependencies, where sub points represent hosting options of the dependencies:
```
Comment (HonkingGoose) [inline lib/manager/terraform/readme.md:1]:
```suggestion
Currently, Terraform supports renovating the following dependencies, where sub points represent hosting options of the dependencies:
```

I think it reads better with the article `the` added here.

File: lib/manager/terraform/readme.md [hunk 3]
```diff
@@ -1,19 +1,36 @@
...
-```
-
-The following _range_ constraints are also supported:
+Terraform range constraints are supported:
 
 - `>= 1.2.0`: version 1.2.0 or newer
 - `<= 1.2.0`: version 1.2.0 or older
 - `~> 1.2.0`: any non-beta version >= 1.2.0 and < 1.3.0, e.g. 1.2.X
 - `~> 1.2`: any non-beta version >= 1.2.0 and < 2.0.0, e.g. 1.X.Y
 - `>= 1.0.0`, <= 2.0.0`: any version between 1.0.0 and 2.0.0 inclusive
```
Comment (HonkingGoose) [inline lib/manager/terraform/readme.md:23]: The text after the range option is blue, I think this means that something is going wrong with the ` characters (which control whether or not it uses a mono-spaced font and grey background).

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/manager/terraform/readme.md (hunk 3 above)
Comment (secustor) [reply lib/manager/terraform/readme.md:23]: Thx for the suggestions! I have missed those.

File: lib/manager/terraform/readme.md [hunk 4]
```diff
@@ -1,19 +1,36 @@
...
-
-The following _range_ constraints are also supported:
+Terraform range constraints are supported:
 
 - `>= 1.2.0`: version 1.2.0 or newer
 - `<= 1.2.0`: version 1.2.0 or older
 - `~> 1.2.0`: any non-beta version >= 1.2.0 and < 1.3.0, e.g. 1.2.X
 - `~> 1.2`: any non-beta version >= 1.2.0 and < 2.0.0, e.g. 1.X.Y
-- `>= 1.0.0`, <= 2.0.0`: any version between 1.0.0 and 2.0.0 inclusive
+- `>= 1.0.0`, `<= 2.0.0`: any version between 1.0.0 and 2.0.0 inclusive
```
Comment (HonkingGoose) [inline lib/manager/terraform/readme.md:23]:
```suggestion
- `>= 1.0.0, <= 2.0.0`: any version between 1.0.0 and 2.0.0 inclusive
```

I'm not sure if my previous suggestion was correct, maybe it should be this instead???

`>= 1.0.0, <= 2.0.0`

File: lib/manager/terraform/readme.md [hunk 5]
```diff
@@ -1,19 +1,36 @@
-Currently Terraform support is limited to Terraform registry sources and github sources that include SemVer refs, e.g. like `github.com/hashicorp/example?ref=v1.0.0`.
+Currently, Terraform supports renovating the following dependencies, where sub points represent hosting options of the dependencies:
```
Comment (HonkingGoose) [inline lib/manager/terraform/readme.md:1]:
```suggestion
Terraform supports renovating the following dependencies, where sub points represent hosting options of the dependencies:
```

To me the use of the words "Currently, " implies that on a later date Renovate/Terraform will support more renovations. By not using such words as "currently, later, right now, in future" we don't create expectations of further improvements. The sentence would read just as nice if we removed the word "Currently, ".

Also by not using words like "Currently" we free ourselves from updating the docs when at a later time "Currently, Terraform supports..." some other renovation as well.

So overall, I think using "time-based" language like this is a bit dangerous.

But let's see what the rest of the Renovate team thinks about using words like this in our docs... 
So don't rush into applying this suggestion right away, as I'm curious what the rest of the Renovate team has to say. Maybe I'm totally honking up the wrong pond here... 😄

**Comment by rarkins** [comment]
FYI I'm marking all breaking change PRs as draft to reduce the chance of accidental merge

**Review by rarkins — DISMISSED** [review]

**Comment by rarkins** [comment]
@secustor I seem to have broken this in the merge. Could you take a look?

**Comment by secustor** [comment]
@rarkins  Seems like a new test has been added, fixed the snapshot.

**Pull request merged**